.github/workflows/build.yml

```diff
@@ -14,52 +14,49 @@ concurrency:
 
 jobs:
   haskell:
-    runs-on: ${{ matrix.os }}
     strategy:
+      fail-fast: true
       matrix:
         os:
           - macos-latest
           - ubuntu-latest
-        ghc:
-          - "9.0"
-          - "8.10"
+        devShell:
+          - ghc810
+          - ghc90
+
+    runs-on: ${{ matrix.os }}
 
     steps:
-      - name: Checkout
-        uses: actions/checkout@v3
-      - name: Install dependencies (Linux)
-        if: ${{ runner.os == 'Linux' }}
-        run: |
-          sudo apt-get update
-          sudo apt-get install zsh
-      - name: Install dependencies (macOS)
-        if: ${{ runner.os == 'macOS' }}
-        run: |
-          brew update
-          brew install zsh
-      - name: Install Haskell tooling
-        uses: haskell/actions/setup@v2
+      - uses: actions/checkout@v3
+      - uses: cachix/install-nix-action@v18
         with:
-          ghc-version: ${{ matrix.ghc }}
-          cabal-version: "3.6"
-      - name: Configure project
-        run: cabal configure --enable-tests
-      - name: Freeze dependencies
-        run: cabal freeze
-      - name: Cache dependencies
-        uses: actions/cache@v3
+          # https://input-output-hk.github.io/haskell.nix/tutorials/getting-started.html
+          extra_nix_config: |
+            trusted-public-keys = hydra.iohk.io:f/Ea+s+dFdN+3Y/G+FDgSq+a5NEWhJGzdjvKNGv0/EQ= cache.nixos.org-1:6NCHdD59X431o0gWypbMrAURkbJ16ZPMQFGspcDShjY=
+            substituters = https://cache.iog.io https://cache.nixos.org
+      # https://github.com/actions/cache/blob/main/examples.md#haskell---cabal
+      - uses: actions/cache@v3
         with:
           path: |
+            ~/.cabal/packages
             ~/.cabal/store
             dist-newstyle
-          key: ${{ runner.os }}-${{ matrix.ghc }}-${{ hashFiles('cabal.project.freeze') }}
-          restore-keys: ${{ runner.os }}-${{ matrix.ghc }}-
-      - name: Build project
-        run: cabal build
+          key: ${{ runner.os }}-${{ matrix.devShell }}-${{ hashFiles('*.cabal') }}
+          restore-keys: ${{ runner.os }}-${{ matrix.devShell }}-
+      - name: Install packages / Print versions
+        run: ./bin/${{ matrix.devShell }} -c ghc --version && cabal --version
+      - name: Update package index
+        run: ./bin/${{ matrix.devShell }} -c cabal update
+      - name: Enable tests
+        run: ./bin/${{ matrix.devShell }} -c cabal configure --enable-tests
+      - name: Install project dependencies
+        run: ./bin/${{ matrix.devShell }} -c cabal build --only-dependencies
+      - name: Compile project
+        run: ./bin/${{ matrix.devShell }} -c cabal build
       - name: Run tests
-        run: cabal test
+        run: ./bin/${{ matrix.devShell }} -c cabal test
       - name: Check documentation
-        run: ./script/haddock
+        run: ./bin/${{ matrix.devShell }} -c ./script/haddock
```

Comment:
@CristhianMotoche I see your point, however, I think it would complicate troubleshooting a little bit since it adds a level of indirection. What do you think about refining this as we find more use cases on the go?

Reply:
Sounds good to me. 👍